src/audio/dp_queue.c

```diff
@@ -0,0 +1,328 @@
+
+	uint32_t write_offset;
+	uint32_t read_offset;
+	uint32_t available_data; /* amount of data ready for immediate reading */
+	/* free buffer space
+	 * NOTE!
+	 *  - when dp queue is shared, available_data + free_space DOES NOT eq data_buffer_size
+	 *  - when dp queue is not shared available_data + free_space always == data_buffer_size
+	 */
+	uint32_t free_space;
+
+	struct sof_audio_stream_params audio_stream_params;
+
+	struct k_spinlock lock;
+};
+
+static inline bool dp_queue_is_shared(struct dp_queue *dp_queue)
+{
+	return !!(dp_queue->flags & DP_QUEUE_MODE_SHARED);
+}
+
+static inline uint8_t *dp_queue_buffer_end(struct dp_queue *dp_queue)
+{
+	return dp_queue->data_buffer + dp_queue->data_buffer_size;
+}
+
+struct dp_queue *dp_queue_get_from_sink(struct sof_sink __sparse_cache *sink)
+{
+	struct dp_queue_sink_source_c __sparse_cache *sink_src_c =
+		attr_container_of(sink, struct dp_queue_sink_source_c __sparse_cache,
+				  sink, __sparse_cache);
+
+	return sink_src_c->owner;
+}
+
+struct dp_queue *dp_queue_get_from_source(struct sof_source __sparse_cache *source)
+{
+	struct dp_queue_sink_source_c __sparse_cache *sink_src_c =
+		attr_container_of(source, struct dp_queue_sink_source_c __sparse_cache,
+				  source, __sparse_cache);
+	return sink_src_c->owner;
+}
+
+static inline void dp_queue_invalidate_shared(struct dp_queue *dp_queue,
+					      void __sparse_cache *ptr, uint32_t size)
+{
+	/* rollback? */
+	if ((uintptr_t)ptr + size > (uintptr_t)dp_queue_buffer_end(dp_queue)) {
+		/* writeback till the end of circular buffer */
+		dcache_invalidate_region
+			(ptr, (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr);
+		size -= (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr;
+	}
+	/* invalidate rest of data - from begin of buffer, remaining size */
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)dp_queue->data_buffer, size);
+}
+
+static inline k_spinlock_key_t dp_queue_lock(struct dp_queue *dp_queue)
+{
+	return k_spin_lock(&dp_queue->lock);
+}
+
+static inline void dp_queue_unlock(struct dp_queue *dp_queue, k_spinlock_key_t key)
+{
+	k_spin_unlock(&dp_queue->lock, key);
+}
+
+struct sof_sink __sparse_cache *dp_queue_get_sink(struct dp_queue *dp_queue)
+{
+	return &dp_queue->sink_src_c->sink;
+}
+
+struct sof_source __sparse_cache *dp_queue_get_source(struct dp_queue *dp_queue)
+{
+	return &dp_queue->sink_src_c->source;
+}
+
+static size_t dp_queue_get_free_size(struct sof_sink __sparse_cache *sink)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	return dp_queue->free_space;
+}
+
+static int dp_queue_get_buffer(struct sof_sink __sparse_cache *sink, size_t req_size,
+			       void **data_ptr, void **buffer_start, size_t *buffer_size)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	/* dp_queue_get_free_size will return free size with adjustment for cacheline if needed */
+	if (req_size > dp_queue->ibs || req_size > dp_queue_get_free_size(sink))
```

Comment:
isn't free size never greater than `.ibs`? If so, just the second half of the check is enough